ansys/mapdl/core/license.py

```diff
@@ -0,0 +1,349 @@
+"""
+Everything related to licensing and license serve checks. 
+
+"""
+
+import os 
+import time 
+import warnings
+import re 
+import socket 
+import subprocess
+import datetime
+
+
+from ansys.mapdl.core.errors import LicenseServerConnectionError  
+
+
+LOCALHOST = '127.0.0.1'
+
+
+def try_license_file():
+    
+    try:
+        license_file_checker()
+    except LicenseServerConnectionError:
+        # expected error, so let it raise
+        raise 
+    except:
+        # Rest of the cases, let's run secondary methods.
+        # print("Error found trying to read license file.")
+        return False
+    else:
+        return True 
+
+
+def license_file_checker():
+    licdebug_file = os.path.join(get_licdebug_path(), get_licdebug_name())
+    file_iterator = get_licdebug_msg(licdebug_file)
+
+    time_to_be_checking = 10 #seconds
+    max_time = datetime.time(0, 0, time_to_be_checking) #hours, minute, seconds
+
+    while max_time < datetime.datetime.now().time():
+        msg = next(file_iterator)
+
+        if is_denied_msg(msg):
+            license_path = re.findall( "(?<=License path:)(.*)(?=;\n)", msg)[0]
+            license_port = license_path.split('@')[0]
+            license_hostname = license_path.split('@')[1]
+            raise LicenseServerConnectionError(
+                head_message=f"Error connecting to {license_port}:{license_hostname}",
+                error_message=msg,
+                tail_message=f"Error found in file {get_licdebug_name()}")
+        
+    else:
+        print("Time is out for license checking.")
+
+def get_licdebug_path():
+    if os.name == 'nt': 
+        folder = os.getenv('TEMP')
+    elif os.name == 'posix':
+        folder = os.getenv('HOME')
+    else:
+        raise OSError
+    
+    return os.path.join(folder, '.ansys')
+
+
+def get_licdebug_name():
+    # Licdebug name convention:
+    # - For version 22.1 and above: `licdebug.$hostname.$appname.$version.out`
+    # - For version 21.2 and below: `licdebug.$appname.$version.out`
+
+    from ansys.mapdl.core.launcher import _version_from_path, get_ansys_path 
```

Comment:
Is there a specific reason this is not imported at the top of the file?

Reply:
None. Feel free to move it up.

Reply:
```suggestion
```

Reply:
I would move it to line 7:

`from ansys.mapdl.core.launcher import _version_from_path, get_ansys_path`

Reply:
wait a minute.... I jsut remembered. I gave me 'circular import' error. That's why those import are there. Don't you have that error too? (Maybe you are not running it)

Reply:
I would suggest not to import `_version_from_path, get_ansys_path` from launcher, since you use the license methods in launcher. You only use these two methods to get `version` on line 74: 
`version = _version_from_path(get_ansys_path(allow_input=False))`
Instead, I would get `version` inside of launcher.py before you make the call to `try_license_server` and add the version as an argument to this method. This should remove the circular dependency.

Reply:
Issue comes down to how we get our version.  Do we just specify the version of MAPDL we want to launch and hope that the user has MAPDL installed in the right location (with the right env var), or do we allow the specification of the executable file?

If we go with specifying the version, do we allow the ability to override the exec_file?  I know that PyAEDT does that (allows for version specification).  We could make it consistent here and make it the primary way of launching MAPDL, permitting `exec_file` for those who want to specify a custom file.

If we do this, let's do this in a follow up PR.